mm2src/mm2_main/src/lp_healthcheck.rs

```diff
@@ -0,0 +1,444 @@
+    static ADDRESS_RECORD_EXPIRATION: OnceLock<Duration> = OnceLock::new();
+
+    let address_record_exp =
+        ADDRESS_RECORD_EXPIRATION.get_or_init(|| Duration::from_secs(ctx.healthcheck.config.timeout_secs));
+
+    let target_peer_id = req.peer_id;
+
+    let p2p_ctx = P2PContext::fetch_from_mm_arc(&ctx);
+    if target_peer_id == p2p_ctx.peer_id() {
+        // That's us, so return true.
+        return Ok(true);
+    }
+
+    let message =
+        HealthcheckMessage::generate_message(&ctx, target_peer_id, false, ctx.healthcheck.config.message_expiration)
+            .map_err(|reason| HealthcheckRpcError::MessageGenerationFailed { reason })?;
+
+    let encoded_message = message
+        .encode()
+        .map_err(|e| HealthcheckRpcError::MessageEncodingFailed { reason: e.to_string() })?;
+
+    let (tx, rx): (Sender<()>, Receiver<()>) = oneshot::channel();
+
+    let mut book = ctx.healthcheck.response_handler.lock().await;
+    book.clear_expired_entries();
+    book.insert(target_peer_id.to_string(), tx, *address_record_exp);
+    drop(book);
+
+    broadcast_p2p_msg(&ctx, peer_healthcheck_topic(&target_peer_id), encoded_message, None);
+
+    let timeout_duration = Duration::from_secs(ctx.healthcheck.config.timeout_secs);
+    Ok(rx.timeout(timeout_duration).await == Ok(Ok(())))
+}
+
+pub(crate) async fn process_p2p_healthcheck_message(ctx: &MmArc, message: mm2_libp2p::GossipsubMessage) {
+    macro_rules! try_or_return {
+        ($exp:expr, $msg: expr) => {
+            match $exp {
+                Ok(t) => t,
+                Err(e) => {
+                    log::error!("{}, error: {e:?}", $msg);
+                    return;
+                },
+            }
+        };
+    }
+
+    let data = try_or_return!(
+        HealthcheckMessage::decode(&message.data),
+        "Couldn't decode healthcheck message"
+    );
+
+    let sender_peer = data.sender_peer().to_owned();
+
+    let mut ddos_shield = ctx.healthcheck.ddos_shield.lock().await;
+    ddos_shield.clear_expired_entries();
+    if ddos_shield
+        .insert(
+            sender_peer.to_string(),
+            (),
+            Duration::from_millis(ctx.healthcheck.config.blocking_ms_for_per_address),
+        )
+        .is_some()
+    {
+        log::warn!("Peer '{sender_peer}' exceeded the healthcheck blocking time, skipping their message.");
+        return;
+    }
+    drop(ddos_shield);
+
+    let ctx_c = ctx.clone();
+
+    // Pass the remaining work to another thread to free up this one as soon as possible,
+    // so KDF can handle a high amount of healthcheck messages more efficiently.
+    ctx.spawner().spawn(async move {
+        let my_peer_id = P2PContext::fetch_from_mm_arc(&ctx_c).peer_id();
+        if !data.is_received_message_valid(my_peer_id) {
+            log::error!("Received an invalid healthcheck message.");
+            log::debug!("Message context: {:?}", data);
+            return;
+        };
```

Comment:
nit: I would go for doing this check outside the spawned task to avoid spawning unnecessary tasks.

Reply:
The reason for spawning another thread is to make the current thread available as quickly as possible so it’s ready to handle the next message.